### PR TITLE
Fix node equality check in embedding processor

### PR DESCRIPTION
**Pull request by justinas**
The current comparison "by name" is broken, since `node.GetName()` returns just the node "name" (UUID), while embedding's `GetEmbeddedID()` returns `node/{uuid}`:

https://github.com/gravitational/teleport/blob/5fcc1ee2b48b4002b207eb5398e64de81e0869d3/lib/ai/embedding/embedding.go#L69-L73

As such, we currently never match nodes with their existing embeddings and wastefully recompute embeddings on every iteration.

### CODE REVIEW

**Review by hugoShaka — APPROVED** [review]

File: lib/ai/embeddingprocessor_test.go [hunk 1]
```diff
@@ -141,6 +163,18 @@ func TestMarshallUnmarshallEmbedding(t *testing.T) {
...
+	node, _ := types.NewServer(fmt.Sprintf("node%d", index), types.KindNode, types.ServerSpecV2{
+		Addr:     "127.0.0.1:1234",
+		Hostname: fmt.Sprintf("node%d", index),
+		CmdLabels: map[string]types.CommandLabelV2{
+			"version":  {Result: "v8"},
+			"hostname": {Result: fmt.Sprintf("node%d.example.com", index)},
+		},
+	})
+	return node
+}
```
Comment (codingllama) [inline lib/ai/embeddingprocessor_test.go:176]: nit:

```suggestion
func makeNode(num int) types.Server {
	node, _ := types.NewServer(fmt.Sprintf("node%d", num), types.KindNode, types.ServerSpecV2{
		Addr:     "127.0.0.1:1234",
		Hostname: fmt.Sprintf("node%d", num),
		CmdLabels: map[string]types.CommandLabelV2{
			"version":  {Result: "v8"},
			"hostname": {Result: fmt.Sprintf("node%d.example.com", num)},
		},
	})
	return node
}
```

File: lib/ai/embeddingprocessor_test.go [hunk 2]
```diff
@@ -123,6 +123,28 @@ func TestNodeEmbeddingGeneration(t *testing.T) {
...
+
+	// Run once more and verify that only changed nodes get their embeddings recalculated
+	node1 := nodes[0]
+	node1.GetMetadata().Labels["foo"] = "bar"
+	_, err = presence.UpsertNode(ctx, node1)
+	require.NoError(t, err)
+
+	ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+	processor.Process(ctx)
```
Comment (codingllama) [inline lib/ai/embeddingprocessor_test.go:139]: Is Process async? Why do we need the cancelable context here?

**Review by codingllama — COMMENTED** [review]
Well spotted.

File: lib/ai/embeddingprocessor_test.go (hunk 2 above)
Comment (justinas) [reply lib/ai/embeddingprocessor_test.go:139]: Good point, we do not!

File: lib/ai/embeddingprocessor.go [hunk 3]
```diff
@@ -201,14 +201,16 @@ func (e *EmbeddingProcessor) Run(ctx context.Context, initialDelay, period time.
...
-			e.process(ctx)
+			e.Process(ctx)
 		}
 	}
 }
 
-func (e *EmbeddingProcessor) process(ctx context.Context) {
+// Process updates embeddings for all nodes once. It is only exposed for use in tests,
+// the service should use Run() to do this periodically instead.
+func (e *EmbeddingProcessor) Process(ctx context.Context) {
```
[the remark this answers is not in the document]
Comment (justinas) [reply lib/ai/embeddingprocessor.go]: Changed up the test to only use the public API (`Run()`).

File: lib/ai/embeddingprocessor_test.go [hunk 4]
```diff
@@ -86,25 +93,16 @@ func TestNodeEmbeddingGeneration(t *testing.T) {
 		Jitter:              retryutils.NewSeventhJitter(),
 	})
 
-	done := make(chan struct{})
```
Comment (justinas) [inline lib/ai/embeddingprocessor_test.go:89]: @hugoShaka I've removed this channel now that we run the processor throughout the test (and we want to observe changes after modifying nodes). IMO it should be alright to query the calculated embeddings although (no-op) updates might be occurring in the background? The race detector doesn't complain.

Comment (hugoShaka) [reply lib/ai/embeddingprocessor_test.go:89]: Yeah, this is very likely a remain from my initial asynchronous implementation. I replaced it with Jakub's one and never adapted the test.

File: lib/ai/embeddingprocessor.go [hunk 5]
```diff
@@ -208,6 +208,7 @@ func (e *EmbeddingProcessor) Run(ctx context.Context, initialDelay, period time.
 	}
 }
 
+// Process updates embeddings for all nodes once.
```
Comment (codingllama) [inline lib/ai/embeddingprocessor.go:211]:
```suggestion
// process updates embeddings for all nodes once.
```

**Review by codingllama — APPROVED** [review]

**Review by jakule — APPROVED** [review]

**Comment by justinas** [comment]
> @justinas See the table below for backport results.
> Branch 	Result
> branch/v13 	[Failed](https://github.com/gravitational/teleport/runs/15939244972?check_suite_focus=true)

Should backport cleanly once https://github.com/gravitational/teleport/pull/30495 merges

**Pull request merged**